### PR TITLE
feature: Prepare extraction of main bucket

### PR DESCRIPTION
**Pull request by Ash258**
#2939, #2609, #1328

❗❗❗ This PR is not meant to be merged. ❗❗❗

I want to only proof, that extraction of bucket is possible and pretty easy. Bucket folder will not be removed from this PR (to keep diff cleaner, without millions of lines deleted due to bucket folder deletion). If collaborators (This should be done by @r15ch13, since he has access to Excavator) decided to adopt this change, just follow steps bellow ([co-authored-by tag would be nice for all commits / PR 😍](https://github.blog/2018-01-29-commit-together-with-co-authors/)):

- \* SideNote: I will push small tweaks, which i found, so noone should fix them inside bucket extraction phase (like lib-exec bucket moving, don't use scoop directly inside update functions, update branch reflection)
    - ec524cf0688d6af3b9ffbe6838337bfb3a7d5a3d - #3061
    - fe873182549464ed77fe90562bc3614489074ebe - #3062
    - 147d49b3ebb1a70ed524a5cf97be8c163c2d0e31 - #3063

## How to?

Slighlty edited and step by step flow from https://github.com/lukesampson/scoop/issues/1328#issuecomment-458355840

1. [x] Postpone excavator
1. [x] Extract bucket folder with history
	- `git filter-branch --prune-empty --subdirectory-filter bucket`
		- Do it inside normal terminal (not vscode)
			- There will be lots of changes and vscode could be unstable
		- It takes quite a while
		- For me it was ~50 minutes
		- Time
	- `git remote set-url origin 'https://github.com/ScoopInstaller/scoop-main'`
	- `git push -u origin`
	- New bucket repository is DONE
		- Sure, there is need to be setuped readme, appveyor, format, binaries, contributing and all other bucket repository things
1. [x] Create new branch (bucket-removal-adoption)
	- All changes to reflect this change
	- Maybe even PR for some discussion❓
1. [x] buckets.json
	- Add main bucket into known buckets
1. [x] Fix code
	1. [x] Install.ps1
		- Download main bucket while installing
			- Since git, 7zip are needed
			- Any better way?
		- checkup should look for added bucket
		- `scoop update`
			- Current users could not even notice change
			- While updating, check for bucket
				- If there is not added, add it
	1. [x] General code
	1. [x] binaries default dir parameter
1. [ ] Move all package requests (not core related) issues into new repository
1. [ ] Delete bucket folder from core
1. [ ] Run beautifier / vscode format document on all files inside core repository
	1. [ ] Use [`UseCorrectCasing`](https://github.com/PowerShell/PSScriptAnalyzer/blob/development/RuleDocumentation/UseCorrectCasing.md) rule
1. [ ] Make master protected
1. [ ] Create develop branch
	- Notify users, who want to test earlier changes
		- `scoop config SCOOP_BRANCH develop`
			- There is need little fix for this.
			- When you change it, you need to reset to that branch.
				- No it's needed to do it manually using command line.
			- Basicly if `git branch -ne get-config SCOOP_BRANCH { git checkout get-config SCOOP_BRANCH }`
		- Create issue, about how to use develop branch and what does it involve
		- Pin that issue
	- Never push to master directly
		- All changes will happen in develop branch
		- When change is done, debugged and fully working, PR will be created to master with description like `Version: X.XX.XX Adopt new change`
1. [x] Change excavator config with new repository
1. [x] Start excavator again
1. [ ] ❓ Rename this bucket into `Core` or something
1. [ ] ❓ Move it to same organization
1. [ ] ❓ Tag and release version 1.0
    - Scoop could adopt normal release process with release tab and changelog inside it.
1. [ ] 🎉🎉 Welcome new era of Scoop with easier maintaining and expanding 🎉🎉

### CODE REVIEW

**Comment by Ash258** [comment]
Updating works even after this change. 

If there is no `bucket` property inside install.json, 'main' is considered. So updates will be OK, without needing to reinstall.

Only inside scoop list will be `[main]` instead of empty collumn.

**Comment by Ash258** [comment]
How to test?

1. docker-compose up --build -d; docker-compose run scoop
1. `$env:SCOOP = 'SOME NEW PATH TO TEST'`
1. `PS: bin> .\install.ps1`
1. `Rename-Item "$env:SCOOP\apps\scoop\current\bucket" "new-bucket"`
1. Test installing, uninstall, whatever

**Comment by Ash258** [comment]
I haven't touched on tests yet. But after recent changes (@r15ch13, 589303facc5284f6f95c1305191e0558c0169691), it should be fine.

**Comment by Ash258** [comment]
Feel free to suggest any change.

/fyi @r15ch13 @rasa @h404bi

**Comment by h404bi** [comment]
There is one thing **VERY important** that you didn't cover above. We need to handle those previously installed packages so that users can still update these packages after this breaking change.

packages installed from the `extras` bucket has a bucket mark.
**install.json**
```json
{
    "bucket": "extras",
    "architecture": "64bit"
}
```

but packages installed from `core` bucket doesn't.
**install.json**
```json
{
    "architecture": "64bit"
}
```

**Comment by Ash258** [comment]
It's already handled :) 

2f16f0d1370b9cc50534fd3407a949e2e8377f27

**Comment by niheaven** [comment]
Like brew? Coooool!

**Comment by h404bi** [comment]
If a user installs scoop from the first time, the installer script downloads a zip of the core bucket (this is a normal logic, to make scoop work at the first time, git and 7zip are needed). When the user runs `scoop update` for the first time, they will be notified to install git and 7zip. 

But the `zip version` of the core bucket is not a git repository after the installation. So there should be a handler logic to clone a new core bucket and remove the zip one in `scoop-update.ps1`.

**Comment by h404bi** [comment]
@Ash258 make sense.

**Comment by Ash258** [comment]
Ok. Everything is no prepared. This could be even merged, since all references were changed.

Today / tomorrow when I arrive into flat I will run some docker testing with updating already installed scoop and clean installation, binaries.

When i do not find any bugs this could be even merged. There are still few steps needed by @r15ch13, regarding excavator settings, but this PR now ready to merge. (Best possible to merge it into develop branch, where it will be prepared and then PR it into master, so users should not even notice some problems.)

@h404bi, @niheaven, @r15ch13 Pleaes review suggest alternative to whatever.

File: bin/install.ps1 [hunk 1]
```diff
@@ -42,23 +42,35 @@ $dir = ensure (versiondir 'scoop' 'current')
...
+
+Write-Output 'Extracting...'
+unzip $zipfile "$dir\_tmp"
+Copy-Item "$dir\_tmp\*-master\*" $dir -r -force
+Remove-Item "$dir\_tmp", $zipfile -r -force
+
 ensure_robocopy_in_path
 ensure_scoop_in_path
-scoop config lastupdate ([System.DateTime]::Now.ToString('o'))
-success 'Scoop was installed successfully!'
```
Comment (h404bi) [inline bin/install.ps1]: keep the successful message?

File: lib/buckets.ps1 [hunk 2]
```diff
@@ -2,26 +2,36 @@
...
+    param(
+        [string] $Name = 'main',
+        [switch] $Root
+    )
 
-    if ($name) {
-        $bucket = "$bucketsdir\$name"
-    }
-    if (Test-Path "$bucket\bucket") {
+    $bucket = "$bucketsdir\$Name" # main bucket
```
Comment (h404bi) [inline lib/buckets.ps1]: The comment `main bucket` may be `main bucket by default`

File: libexec/scoop-install.ps1 [hunk 3]
```diff
@@ -4,12 +4,7 @@
 #      scoop install git
 #
 # To install an app from a manifest at a URL:
-#      scoop install https://raw.githubusercontent.com/lukesampson/scoop/master/bucket/runat.json
-#
-# To install an app from a manifest on your computer
-#      scoop install \path\to\app.json
```
Comment (h404bi) [inline libexec/scoop-install.ps1]: Keep this tip?

File: buckets.json [hunk 4]
```diff
@@ -1,5 +1,6 @@
 {
-    "extras": "https://github.com/lukesampson/scoop-extras",
+    "main": "https://github.com/scoopinstaller/scoop-main",
+    "extras": "https://github.com/lukesampson/scoop-extras.git",
```
Comment (h404bi) [inline buckets.json]: trim `.git` suffix?

File: lib/autoupdate.ps1 [hunk 5]
```diff
@@ -141,15 +139,6 @@ function find_hash_in_headers([String] $url) {
 function get_hash_for_app([String] $app, $config, [String] $version, [String] $url, [Hashtable] $substitutions) {
     $hash = $null
 
-    <#
-    TODO implement more hashing types
```
[the remark this answers is not in the document]
Comment (Ash258) [reply lib/autoupdate.ps1]: All are already implemented

[the remark this answers is not in the document]
Comment (niheaven) [reply lib/autoupdate.ps1]: I mean `TODO` more, and leave these implemented alone as comment. Maybe need to rearrange these lines.

File: bin/auto-pr.ps1 [hunk 6]
```diff
@@ -27,22 +27,23 @@
     PS REPODIR > .\bin\auto-pr.ps1 -Push
```
[the remark this answers is not in the document]
Comment (Ash258) [reply bin/auto-pr.ps1]: After implementation of nested bucket folder I would say BUCKETROOT would be more reliable

[the remark this answers is not in the document]
Comment (Ash258) [reply bin/auto-pr.ps1]: Changed to BUCKETROOT

**Comment by Ash258** [comment]
Made `$Dir` Parameter inside binaries mandatory and defaults are gone.

**Comment by Ash258** [comment]
I think this is ready to be merged into develop branch. We can test it today / tomorrow in develop branch and then merge it to master

**Comment by r15ch13** [comment]
@Ash258 can you squash some of the commits? Or should I squash the whole PR?

**Comment by Ash258** [comment]
Squash whole PR.

**Comment by Ash258** [comment]
Looks good after fix for switch branches ❤

A

**Pull request merged**